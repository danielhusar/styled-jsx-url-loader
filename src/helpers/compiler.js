import path from 'path'
import webpack from 'webpack'
import { createFsFromVolume, Volume } from 'memfs'

export default (fixture, options = {}) => {
  const compiler = webpack({
    context: __dirname,
    entry: fixture,
    output: {
      path: path.resolve(__dirname),
      filename: 'bundle.js',
    },
    module: {
      rules: [
        {
          test: /\.s?css$/,
          use: [
            {
              loader: path.resolve(__dirname, '../loader.js'),
              options,
            },
            {
              loader: require('styled-jsx/webpack').loader,
              options: {
                type: 'scoped',
              },
            },
          ],
        },
      ],
    },
  })

  const volume = createFsFromVolume(new Volume())
  volume.join = path.join
  compiler.outputFileSystem = volume

  return new Promise((resolve, reject) => {
    compiler.run((err, stats) => {
      if (err) reject(err)
      if (stats.hasErrors()) reject(new Error(stats.toJson().errors))
      resolve(stats)
    })
  })
}
