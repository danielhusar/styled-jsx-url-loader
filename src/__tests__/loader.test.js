import path from 'path'
import compiler from '../helpers/compiler.js'

describe('Loader', () => {
  test('Replaces all urls in the css file', async () => {
    const stats = await compiler(path.join(__dirname, './fixtures/styles.css'))
    const output = stats.toJson().modules.slice(-1)[0].source
    expect(output).toMatchSnapshot()
  })

  test('Replaces all urls in the scss file', async () => {
    const stats = await compiler(path.join(__dirname, './fixtures/styles.scss'), { scss: true })
    const output = stats.toJson().modules.slice(-1)[0].source
    expect(output).toMatchSnapshot()
  })

  test('Exclude files', async () => {
    const stats = await compiler(path.join(__dirname, './fixtures/styles.css'), {
      exclude: /image\.png/,
    })
    const output = stats.toJson().modules.slice(-1)[0].source
    expect(output).toMatchSnapshot()
  })

  test('Show warning for invalid scss file', async () => {
    expect.assertions(1)
    console.warn = jest.fn()
    try {
      await compiler(path.join(__dirname, './fixtures/invalid.scss'), { scss: true, debug: true })
    } catch {
      expect(console.warn).toHaveBeenCalledWith(
        'styled-jsx-url-loader: Urls inside variables are not supported',
      )
      console.warn.mockReset()
    }
  })
})
