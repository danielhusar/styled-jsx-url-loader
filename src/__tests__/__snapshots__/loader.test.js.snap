// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader Exclude files 1`] = `
"
import postcssStyledJsxUrlLoader_0 from './font.woff2';
import postcssStyledJsxUrlLoader_1 from './font.woff';
import css from 'styled-jsx/css';

export default css.global\`/* @styled-jsx=global */

header {
  background: lightblue url('./image.png') no-repeat fixed center;
  src: url(\\"\${postcssStyledJsxUrlLoader_0}\\") format('woff2'), url(\\"\${postcssStyledJsxUrlLoader_1}\\") format('woff');
}

footer {
  background: url('./image.png');
}

main {
  background: url(./image.png);
}

div {
  background: url(image.png);
}

.class {
  background: url(./image.png);
}

/* .class {
  background: url(./image.png);
} */

span {
  background: url('https://www.danielhusar.sk/image.png');
}
\`"
`;

exports[`Loader Replaces all urls in the css file 1`] = `
"
import postcssStyledJsxUrlLoader_0 from './image.png';
import postcssStyledJsxUrlLoader_1 from './font.woff2';
import postcssStyledJsxUrlLoader_2 from './font.woff';
import postcssStyledJsxUrlLoader_3 from './image.png';
import postcssStyledJsxUrlLoader_4 from './image.png';
import postcssStyledJsxUrlLoader_5 from './image.png';
import postcssStyledJsxUrlLoader_6 from './image.png';
import css from 'styled-jsx/css';

export default css.global\`/* @styled-jsx=global */

header {
  background: lightblue url(\\"\${postcssStyledJsxUrlLoader_0}\\") no-repeat fixed center;
  src: url(\\"\${postcssStyledJsxUrlLoader_1}\\") format('woff2'), url(\\"\${postcssStyledJsxUrlLoader_2}\\") format('woff');
}

footer {
  background: url(\\"\${postcssStyledJsxUrlLoader_3}\\");
}

main {
  background: url(\\"\${postcssStyledJsxUrlLoader_4}\\");
}

div {
  background: url(\\"\${postcssStyledJsxUrlLoader_5}\\");
}

.class {
  background: url(\\"\${postcssStyledJsxUrlLoader_6}\\");
}

/* .class {
  background: url(./image.png);
} */

span {
  background: url('https://www.danielhusar.sk/image.png');
}
\`"
`;

exports[`Loader Replaces all urls in the sass file 1`] = `
"
import postcssStyledJsxUrlLoader_0 from './image.png';
import css from 'styled-jsx/css';

export default css\`/* @styled-jsx=scoped */

@import 'main';

main {
  div {
    background: url(\\"\${postcssStyledJsxUrlLoader_0}\\");
  }
}
\`"
`;
